ignite/metrics/ssim.py

```diff
@@ -0,0 +1,171 @@
+        >>> y = y_pred * 1.25
+        >>> ssim = SSIM()
+        >>> ssim.update((y_pred, y))
+        >>> ssim.compute()
+        tensor(0.9520)
+    """
+
+    def __init__(
+        self,
+        kernel_size: Sequence[int] = (11, 11),
+        sigma: Sequence[float] = (1.5, 1.5),
+        data_range: float = None,
+        k1: float = 0.01,
+        k2: float = 0.03,
+        output_transform: Callable = lambda x: x,
+    ):
+        if len(kernel_size) != 2 or len(sigma) != 2:
+            raise ValueError(
+                "Expected `kernel_size` and `sigma` to have the length of two."
+                f" Got kernel_size: {len(kernel_size)} and sigma: {len(sigma)}."
+            )
+
+        if any(x % 2 == 0 or x <= 0 for x in kernel_size):
+            raise ValueError(f"Expected `kernel_size` to have odd positive number. Got {kernel_size}.")
+
+        if any(y <= 0 for y in sigma):
+            raise ValueError(f"Expected `sigma` to have positive number. Got {sigma}.")
+
+        self.kernel_size = kernel_size
+        self.sigma = sigma
+        self.data_range = data_range
+        self.k1 = k1
+        self.k2 = k2
+        super(SSIM, self).__init__(output_transform=output_transform)
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_batchwise_ssim = 0.0
+        self._num_examples = 0
+
+    def _gaussian_kernel(self, channel, kernel_size, sigma, device):
+        def gaussian(kernel_size, sigma, device):
+            gauss = torch.arange(
+                start=(1 - kernel_size) / 2, end=(1 + kernel_size) / 2, step=1, dtype=torch.float32, device=device
+            )
+            gauss = torch.exp(-gauss.pow(2) / (2 * pow(sigma, 2)))
+            return (gauss / gauss.sum()).unsqueeze(dim=0)  # (1, kernel_size)
+
+        gaussian_kernel_x = gaussian(kernel_size[0], sigma[0], device)
+        gaussian_kernel_y = gaussian(kernel_size[1], sigma[1], device)
+        kernel = torch.matmul(gaussian_kernel_x.t(), gaussian_kernel_y)  # (kernel_size, 1) * (1, kernel_size)
+
+        return kernel.expand(channel, 1, kernel_size[0], kernel_size[1])
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        y_pred, y = output
+        if y_pred.dtype != y.dtype:
+            raise TypeError(
+                f"Expected `y_pred` and `y` to have the same data type. Got y_pred: {y_pred.dtype} and y: {y.dtype}."
+            )
+
+        if y_pred.shape != y.shape:
+            raise ValueError(
+                f"Expected `y_pred` and `y` to have the same shape. Got y_pred: {y_pred.shape} and y: {y.shape}."
+            )
+
+        if len(y_pred.shape) != 4 or len(y.shape) != 4:
+            raise ValueError(
+                f"Expected `y_pred` and `y` to have BxCxHxW shape. Got y_pred: {y_pred.shape} and y: {y.shape}."
+            )
+
+        if self.data_range is None:
+            self.data_range = max(y_pred.max() - y_pred.min(), y.max() - y.min())
```

Comment:
Does it make sense to setup `data_range` from a single batch ?

Reply:
`data_range` is the maximum difference regardless of batch.
can you suggest more efficient way?

Reply:
Maybe, we should ask user to set it up and forbid to set it as `None` ? In some sense, type of `y` and `y_pred` could determine a default `data_range`.